core/src/util/geom.cpp

```diff
@@ -102,24 +102,27 @@ glm::vec2 worldToScreenSpace(const glm::mat4& _mvp, const glm::vec4& _worldPosit
 }
 
 glm::vec2 centroid(const std::vector<std::vector<glm::vec3>>& _polygon) {
-    glm::vec2 centroid;
-    int n = 0;
 
-    for (auto& l : _polygon) {
-        for (auto& p : l) {
-            centroid.x += p.x;
-            centroid.y += p.y;
-            n++;
-        }
+    if (_polygon.empty()) {
+        return glm::vec2();
     }
 
-    if (n == 0) {
-        return centroid;
-    }
+    // We will consider only the outer ring of the polygon when calculating
+    // the centroid - holes will be ignored. It would be possible to transform
+    // a polygon with holes into a polygon that covers the same set of points
+    // without any holes, but that seems overly complex for our needs.
+    const auto& ring = _polygon.front();
+    glm::vec2 centroid;
+    float area = 0.f;
 
-    centroid /= n;
+    for (auto curr = ring.begin(), prev = ring.end() - 1; curr != ring.end(); prev = curr, ++curr) {
+        float a = (prev->x * curr->y - curr->x * prev->y);
+        centroid.x += (prev->x + curr->x) * a;
+        centroid.y += (prev->y + curr->y) * a;
+        area += a;
+    }
 
-    return centroid;
+    return centroid / (3.f * area);
```

Comment:
Correct, I thought about naming this differently but I couldn't easily express what this represented haha